Add support for calculating GPS Lag over DroneCAN

bugobliterator opened the pull request:
* Adds support for GlobalTimeSync to precisely synchronize GPS Time.
* Adds support for calculating GPS Lag.

Following image shows the calculation of GPS message rx lag.

![image](https://github.com/ArduPilot/ardupilot/assets/5697288/b354a19b-54b6-4031-a5ed-85889cfa8210)

tridge reviewed (changes requested):
I'm pretty sure this will fail for several cases:
 - a GPS module that doesn't have the leap seconds yet when we first get GPS lock and lockin the RTC base time 
 - a large crystal drift leading to large drift between GNSS time and AP_HAL::micros64()

The fundamental problem is the assumption that the two time domains are equivalent

bugobliterator commented:
Explanation of procedure on how the Lag is being Calculated:

Key Time Domains:
**Periph Time Domain** : Time from micros on AP_Periph
**GPS Time Domain**: Time from last_message_epoch_usec() or time_epoch_usec(), both these are calculated using PPS signal.
**Ardupilot Time Domain**: Time from micros on Ardupilot

On Periph Side for Time Synchronisation Message:

* _**DroneCAN Transmit Timestamp[Periph Time Domain]**_: 
On Tx interrupt time is recorded by message id. Tx Interrupt is only called once the transmission is successfully placed on the wire and ACK is received. This time is recorded in AP_Periph local timescale.

* _**Ublox Message Timestamp[GPS Time Domain]**_: 
This is the time for which the solution is calculated by the GPS module, this time does not fluctuate with leapS as shown below.
This is set using last_message_epoch_usec() == UNIX_OFFSET_MSEC + gps_week * AP_MSEC_PER_WEEK + gps_ms. Its to be noted that that none of these parameters jump arbittrarily once lock is achieved. Concerns regarding leap seconds are not applicable as well. Leap seconds is not added to either gps_week or gps_ms. Its a constant value part of UNIX_OFFSET_MSEC.

* _**Ublox Message Timestamp[Periph Time Domain]**_: 
This is the time for which the solution is calculated by the GPS module, just in Periph Time domain
This is simply set through AP_GPS_Backend with PPS pin timestamp as reference.

Finally:
_**DroneCAN Transmit Timestamp[GPS Time Domain]**_ = **DroneCAN Transmit Timestamp[Periph Time Domain]** + **Ublox Message Timestamp[GPS Time Domain]**  - **Ublox Message Timestamp[Periph Time Domain]**

**Ublox Message Timestamp[GPS Time Domain]**  - **Ublox Message Timestamp[Periph Time Domain]** is the Conversion Time Delta between Periph Local Time and GPS Time. 

On Periph Side for GPS Fix Message and GPS RelPosNed Message:

* _**Ublox Message Timestamp[GPS Time Domain]**_:
Same as above, set using last_message_epoch_usec()


On Ardupilot Side:

* _**TimeSync Correction**_:
Conversion Time Delta between GPS Time Domain and Local Time Domain:
_**DroneCAN Transmit Timestamp[GPS Time Domain]**_ - _**DroneCAN TimeSync Receive Timestamp[Local Time Domain]**_


* _**Current Timestamp[GPS Time Domain]**_:
 _**TimeSync Correction**_ + _**Local Timestamp[Local Time Domain]**_

* _**Total Lag**_:
_**Current Timestamp[GPS Time Domain]**_ - _**Ublox Message Timestamp[GPS Time Domain]**_


Benefit of using Time Synchronisation Message is simply that compared to using `pkt.timestamp.usec[GPS Time Domain]` in message itself which contains timestamp recorded at call time on Periph. Messages are put on queue but may not transmitted for sometime if there's too much higher priority traffic on the bus.
We are getting timestamp[GPS Time Domain] when the message was put on the bus and ACKed which near equivalent to when we record Rx Timestamp[Local Time Domain]. So, we can more precisely synchronise time between GPS module and Ardupilot.
With this type of synchronisation GPS PPS to Ardupilot PPS accuracy of less than 40us was achieved. 

Furthermore:
* I have also added code to ensure we handle missed packets properly using transfer id.
* I have also added sanity check to limit the lag calculation to be with 0.3s.

magicrub commented:
You say "precisely synchronize GPS Time" But what people really want is PPS. Does this help create a path for that?

bugobliterator commented:
> You say "precisely synchronize GPS Time" But what people really want is PPS. Does this help create a path for that?

It extends the PPS through dronecan, yes.

tridge commented:
@bugobliterator how did the tests code with the artificial clock drift?

bugobliterator commented:
Clock drift Estimation using PPM based Time synchronisation over DroneCAN is done by calculating the variation in Delta between GPS Time and Local Time. The result is as follows.

Zero PPM Drift added to micros/millis
![ClockDrift_0PPM_Offset](https://github.com/ArduPilot/ardupilot/assets/5697288/938517c9-b9f0-4485-9da0-6f7a728dfb04)

1000 PPM Drift added to micros/millis
![ClockDrift_1000PPM_Offset](https://github.com/ArduPilot/ardupilot/assets/5697288/c534f5cc-73cd-4738-bc18-7d3d934a9882)

Above plots show clock drift estimation at Global Time Synchronisation rate which is 5Hz/200ms. 1000PPM drift means we will be drifting by 1000us every second or 200us every 200ms. This is correctly estimated as shown in 2nd Image. In the 1st Image the clock drift comes out to be around 4us/200ms or around 20ppm, which seems like a close enough combined figure for combined clock drift of Crystal on Here4 (+/-20ppm) and CubeOrangePlus (+/-20ppm)

bugobliterator commented:
Jitter in sample time as seen by EKF after estimated lag is used to calculate the actual GPS Time in Local Time Scale. This is done simply by finding delta between subsequent GPS Times as seen by GPS:

![GPS_Time_Jitter](https://github.com/ArduPilot/ardupilot/assets/5697288/d2e89c38-6ce4-431b-a361-4fedb8c27340)

The Jitter was within +/- 1ms. The Plot is of `gps_time_delta_ms` from following code:
https://github.com/ArduPilot/ardupilot/pull/25732/files#diff-8f617a403545351808c8bf54a6d9fc310593075737558480ca2dcc4f12cb4f46R570-R586
